CS497-F15/apigoteway
# Feature/gin auth proxy refit

What:
goteway using gin, more files and packages
JWT signing handled in apigoteway, key pair auto-generated if not existing and saved to file
wildcard proxying to ms-core, only handles auth, GV-User-Ident header contains user id
querying ms-auth for credentials (apigoteway to ms-auth)
POST /auth | json: username, password | Returns json: user_id
POST /refresh | headers: GV-User-Ident | json:refresh_token | Returns status: 200
GET /refresh | headers: GV-User-Ident | Returns json: refresh_token

token routes (client to apigoteway)
POST /auth | json: username, password | Returns json: access_token, refresh_token
POST /refresh | headers: Authorization: BEARER refresh_token | json: access_token | Returns json: access_token, refresh_token

all other authenticated routes (client to apigoteway)
ANY /* | headers: Authorization: BEARER access_token | * | Returns *

Note: OPTIONS doesn't require authentication
Note: On 401 or non-200 status code, you can use refresh token

Why:
more secure, comprehensive api token
more flexible proxying of requests to μservices
must and can

Notes:
happens to simplify ms-auth
actual test cases are yet to be written

How to test:
First, rewrite ms-auth (start new based off task-microservice maybe) to support credentials checking and token storing
hook up api gateway, use postman to get your tokens after having ms-auth running
make authenticated requests after having ms-core running
apigoteway can take command lines arguments -host, -rauth, etc. (see main.go)
write unit or integration tests
